### DISCUSSION TITLE
vips__tiff_openin_source: unable to open source for input

### DISCUSSION THREAD

**Discussion by revathi-murali** [libvips/libvips#3089] [Q&A]
Hi,

When I try convert the tiff image to png, I get the following exception

```
Not a TIFF or MDI file, bad magic number 0 (0x0)\nvips__tiff_openin_source: unable to open source for input
```

Can you let me know if the issue is with the input image? Or Am I missing some pre-processing step?

**Comment by jcupitt** [comment]
Hi, it sounds like you are passing an empty string instead of a tiff file.

**Comment by revathi-murali** [comment]
Nope, the code exists for a longer time and have not changed anything recently in terms of tiff to png conversion.
So I don't think the issue is with the input image passed as empty file.

**Reply by jcupitt** [reply]
But could it be an empty file? Or a file containing a single zero byte? I would save the source to a file and check it.

With this `source.py`:

```python
#!/usr/bin/python3
  
import sys
import pyvips

source = pyvips.Source.new_from_descriptor(sys.stdin.fileno())
image = pyvips.Image.new_from_source(source, "")

target = pyvips.Target.new_to_descriptor(sys.stdout.fileno())
image.write_to_target(target, ".png")
```

I see:

```
$ cat ~/pics/k2.tif | ./source.py | cat >x.png
```

And it seems to work.

**Reply by revathi-murali** [reply]
I have the file in hand and I could run tiffinfo.  Also in the logs, I could see the size of the file (we print the size of the file in our prod logs) which is valid. So it can't be empty file.

**Reply by jcupitt** [reply]
You're giving the loader a file which begins with a zero byte, so I think the error must be in your code somewhere. I can't help more unless I can reproduce the problem.

**Reply by revathi-murali** [reply]
@jcupitt That was just one example. I also see other bad magic number and not just zero. Some other examples

```
bad magic number 65535 (0xffff)\nvips__tiff_openin_source: unable to open source for input
bad magic number 15870 (0x3dfe)\nvips__tiff_openin_source: unable to open source for input
bad magic number 47046 (0xb7c6)\nvips__tiff_openin_source: unable to open source for input
bad magic number 44356 (0xad44)\nvips__tiff_openin_source: unable to open source for input
```

**Reply by jcupitt** [reply]
The magic number is the first few bytes. A TIFF file must start with either `49 49 2A 00` or `4D 4D 00 2A`.

In all these cases, the data you are supplying does not being with the correct magic number, so libvips can't load it. I think your code is probably supplying the wrong data.

Could your source somehow be being reused? If you use one source object to load two different files in two different threads you might see this error.

But I'm only guessing. You will need to show some code that reproduces this error before it can be debugged.

**Reply by revathi-murali** [reply]
@jcupitt Sorry for getting back to this issue late.

When I ran the exiftool against the tiff file, it showed `Little-endian (Intel, II)` for the Exif byte order. Could that be the issue? Do we need to convert to big endian before reading it?

**Reply by jcupitt** [reply]
No, that should be fine.

This code is used **a lot** and works for everyone else, so I think the problem is almost certainly at your end. You are probably not giving it a valid TIFF file, or you are not connecting up the source correctly.

Post some sample code and data that reproduces the problem and someone can help.
